backend/src/nodes/utils/ncnn_model.py

```diff
@@ -0,0 +1,540 @@
+            del self.param_dict[key]
+        except KeyError:
+            pass
+
+    def __contains__(self, item) -> bool:
+        if item in self.param_dict:
+            return True
+        return False
+
+    def __str__(self) -> str:
+        output = ""
+        for v in self.param_dict.values():
+            if isinstance(v.value, list):
+                output += "-233" + v.id.zfill(2) + "="
+            else:
+                output += v.id + "="
+
+            if isinstance(v.value, float) or (isinstance(v.value, np.float32)):  # type: ignore
+                v_str = np.format_float_scientific(v.value, 6, False, exp_digits=2)
+            elif isinstance(v.value, list):
+                v_str = ",".join([str(n) for n in v.value])
+            else:
+                v_str = str(v.value)
+
+            output += v_str + " "
+
+        return output
+
+    def set_op(self, op: str) -> None:
+        self.op = op
+        self.weight_order = param_schema[op]["weightOrder"]
+
+
+class NcnnLayer:
+    def __init__(
+        self,
+        op_type: str = "",
+        name: str = "",
+        num_inputs: int = 0,
+        num_outputs: int = 0,
+        inputs: Union[List[str], None] = None,
+        outputs: Union[List[str], None] = None,
+        params: Union[NcnnParamCollection, None] = None,
+        weight_data: Union[Dict[str, NcnnWeight], None] = None,
+    ):
+        self.op_type: str = op_type
+        self.name: str = name
+        self.num_inputs: int = num_inputs
+        self.num_outputs: int = num_outputs
+        self.inputs: List[str] = [] if inputs is None else inputs
+        self.outputs: List[str] = [] if outputs is None else outputs
+        self.params: NcnnParamCollection = (
+            NcnnParamCollection(op_type) if params is None else params
+        )
+        self.weight_data: Dict[str, NcnnWeight] = (
+            {} if weight_data is None else weight_data
+        )
+
+    def add_param(
+        self, pid: int, value: Union[float, int, List[Union[float, int]]]
+    ) -> None:
+        self.params[pid] = value
+
+    def add_weight(
+        self,
+        data: Union[float, int, np.ndarray, TensorProto],
+        weight_name: str,
+        quantize_tag: bytes = b"",
+    ) -> bytes:
+        if isinstance(data, TensorProto):
+            data_array = onph.to_array(data)
+        elif isinstance(data, float):
+            data_array = np.array(data, np.float32)
+        elif isinstance(data, int):
+            data_array = np.array(data, np.int32)
+        else:
+            data_array = data
+
+        if quantize_tag == DTYPE_FP16:
+            data_array = data_array.astype(np.float16)
+        self.weight_data[weight_name] = NcnnWeight(data_array, quantize_tag)
+
+        return quantize_tag + data_array.tobytes()
+
+
+class NcnnModel:
+    def __init__(
+        self,
+        node_count: int = 0,
+        blob_count: int = 0,
+        param_path: str = "",
+        bin_path: str = "",
+    ) -> None:
```

Comment:
Okay, let's not do overloading ever again.

Sorry about this, but could you please please separate this into a simple constructor and a static factory method `load_from_files` that essentially does this:

```py
        model = NcnnModel(param_path=param_path, bin_path=bin_path)
        with open(bin_path, "rb") as f:
            model.weights_bin = f.read()
```

Then we could do `NcnnModel.load_from_file(param_path, bin_path)` and `ncnn_model = NcnnModel(node_count, blob_count)`, which are both clean APIs.

Reply:
... Isn't this very similar to what I had in the first place, minus reading from bin_path being in the load_model method?

Reply:
Not quite. I suggested a [static factory method](https://stackoverflow.com/questions/929021/what-are-static-factory-methods). Static factory methods are essentially static methods that create an instance of their class. They are pretty much constructors with a name, so we don't have the overloading problem.